### PR TITLE
Resolve edge cases cought by PHPStan level 7

### PR DESCRIPTION
**Pull request by AJenbo**
Type: bugfix / refactoring / documentation update
Breaking change: no

Resolve about half of the issues detected by PHPStan at level 7. This mostly involves handles a few edge cases and improve some PHPDoc. And a few `assert()` to validate what is asserted by peek().

### CODE REVIEW

**Review by kylekatarnls — APPROVED** [review]

**Comment by AJenbo** [comment]
@kylekatarnls thanks, lets merge it once you approve it

File: src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php [hunk 1]
```diff
@@ -8152,15 +8166,15 @@ private function parseIntegerNumberImage($numberRepresentation)
         switch (substr($numberRepresentation, 0, 2)) {
             case '0x':
             case '0X':
-                return hexdec(substr($numberRepresentation, 2));
+                return (int) hexdec(substr($numberRepresentation, 2));
```
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/main/php/PDepend/Source/Language/PHP/AbstractPHPParser.php]: So this one is done. I discovered that `hexdec` is actually less approximative than the runtime conversion made by PHP for `0x` notation. But it's still way better than previous state where all alternative notation for long numbers returned `double(0)`.

File: src/main/php/PDepend/Source/Language/PHP/PHPParserVersion81.php [hunk 2]
```diff
@@ -124,7 +125,9 @@ protected function parseConstructFormalParameterModifiers()
 
         if ($this->tokenizer->peek() === Tokens::T_READONLY) {
             $modifier |= State::IS_READONLY;
-            $this->tokenStack->add($this->tokenizer->next());
+            $next = $this->tokenizer->next();
+            assert($next instanceof Token);
+            $this->tokenStack->add($next);
```
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/main/php/PDepend/Source/Language/PHP/PHPParserVersion81.php]: Applied after merge.

**Comment by kylekatarnls** [comment]
We have coverage decreasing since those edge-cases are not covered by tests.

It would be nice to see if we can actually reproduce (if those are real edge-cases then we could have tests for them), or if they are only theoretical but cannot happen in practice, if so it would be better to handle with `assert()` rather than actual exceptions.

**Comment by AJenbo** [comment]
Yeah, I tried doing my best to deduce if they where possible or not and used assert() in that way. I currently do not have time to work on this, but I'm happy to review it if you have time to do so.

File: src/main/php/PDepend/Source/AST/AbstractASTType.php [hunk 3]
```diff
@@ -399,6 +399,9 @@ public function getTokens()
      */
     public function setTokens(array $tokens, Token $startToken = null)
     {
+        if (!$tokens) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/main/php/PDepend/Source/AST/AbstractASTType.php:404]: This case could happen if one would extend this class or actually manipulate a node, such as manually:

```php
$enum = new ASTEnum('FooBar');
$enum->setTokens([]);
```

But I don't see why someone would do that, and in our context, you always need at least 1 token, for instance an empty enum such as:

```php
enum FooBar{}
```

Would have 4 tokens: `enum`, `FooBar`, `{` and `}`.

Just quitting silently would for instance make this code:

```php
$enum = new ASTEnum('FooBar');
$enum->setCache(new MemoryCacheDriver());
$enum->setTokens(array(new Token(Tokens::T_ENUM, 'enum', 2, 2, 1, 4)));
$enum->setTokens(array());
var_dump($enum->getTokens()[0]->image);
```
Output: `string "enum"`.

So finally, it's maybe better to throw an exception here.

[the remark this answers is not in the document]
Comment (kylekatarnls) [reply src/main/php/PDepend/Source/AST/AbstractASTType.php:405]: I added an InvalidArgumentException on all setTokens method receiving empty list of token.

It will be backward-compatible since they have code such as `end($tokens)->endLine` that would raise a PHP error anyway.

**Review by AJenbo — COMMENTED** [review]
Thanks for updating the pr 🙂

**Review by AJenbo — COMMENTED** [review]
Nice change regarding `getUnexpectedNextTokenException()` :+1: 

... and `requireNextToken()`, I see coverage is now positive and you might even have addressed some more of the level 7 issues then I originally did.

**Comment by kylekatarnls** [comment]
I hope I don't get carried too far away. I introduced some methods for repeated logic and split `getUnexpectedTokenException` because realizing it breaks the PSR, and has unexpected behavior when a `null` value is the value of the passed argument and not the default value because no argument passed.

And now when we need the next token but no possible end-of-file expected, then we can use `requireNextToken()` so `TokenStreamEndException` is taken care of, and the result is guaranteed to be a `Token`.

**Comment by kylekatarnls** [comment]
Oh and I forgot, I also replaced:

```php
$tokenType = $this->tokenizer->peek();

if (...) {
    return ...
} elseif ($tokenType === Tokenizer::T_EOF) {
    throw new TokenStreamEndException($this->tokenizer);
}

throw $this->getUnexpectedNextTokenException();
```

With just:
```php
$tokenType = $this->tokenizer->peek();

if (...) {
    return ...
}

throw $this->getUnexpectedNextTokenException();
```

Because `TokenStreamEndException` is already handled by `getUnexpectedNextTokenException`. This redundant pattern is actually not new.

**Review by kylekatarnls — APPROVED** [review]
Coverage should be improved, but first we should run it with PHP 8.2, it's still on 7.4 which make some coverage missing just because tests are skipped.

So we'll handle in a next PR.

**Review by kylekatarnls — DISMISSED** [review]

**Review by kylekatarnls — APPROVED** [review]
